pkg/odo/cli/service/delete.go

```diff
@@ -26,89 +29,86 @@ var (
 	Delete an existing service`)
 )
 
-// ServiceDeleteOptions encapsulates the options for the odo service delete command
-type ServiceDeleteOptions struct {
+// DeleteOptions encapsulates the options for the odo service delete command
+type DeleteOptions struct {
 	serviceForceDeleteFlag bool
 	serviceName            string
 	*genericclioptions.Context
 	// Context to use when listing service. This will use app and project values from the context
 	componentContext string
-	// choose between Operator Hub and Service Catalog. If true, Operator Hub
-	csvSupport bool
+	// Backend is the service provider backend (Operator Hub or Service Catalog) that was used to create the service
+	Backend ServiceProviderBackend
 }
 
-// NewServiceDeleteOptions creates a new ServiceDeleteOptions instance
-func NewServiceDeleteOptions() *ServiceDeleteOptions {
-	return &ServiceDeleteOptions{}
+// NewDeleteOptions creates a new DeleteOptions instance
+func NewDeleteOptions() *DeleteOptions {
+	return &DeleteOptions{}
 }
 
-// Complete completes ServiceDeleteOptions after they've been created
-func (o *ServiceDeleteOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
-	if o.csvSupport, err = svc.IsCSVSupported(); err != nil {
-		return err
-	} else if o.csvSupport {
-		o.Context, err = genericclioptions.NewDevfileContext(cmd)
+// Complete completes DeleteOptions after they've been created
+func (o *DeleteOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
+	// decide which service backend to use
+	_, _, err = svc.SplitServiceKindName(args[0])
+	if err != nil {
+		// failure to split provided name into two; hence ServiceCatalogBackend
+		o.Backend = NewServiceCatalogBackend()
+
+		err = nil
 	} else {
-		o.Context, err = genericclioptions.NewContext(cmd)
+		// provided name adheres to the format <operator-type>/<crd-name>; hence OperatorBackend
+		o.Backend = NewOperatorBackend()
 	}
```

Comment:
The logic is the same as create.go. Maybe put them in a single place like types.go in this package.

Reply:
But the struct types are different in both cases. One is `DeleteOptions` and the other is `CreateOptions`. How do you suggest I create a "common" function that would set the backend type for both types?

Reply:
We can use the `args[0]` value and decide the backend type. We can also pass the various options like interactive, fromFile etc for deciding the backend type.

Reply:
> We can use the args[0] value and decide the backend type

We rely on this information at only one place and there's barely a few lines of code for it:
```go
_, _, err = svc.SplitServiceKindName(args[0])
if err != nil {
	// failure to split provided name into two; hence ServiceCatalogBackend
	o.Backend = NewServiceCatalogBackend()

	err = nil
}
```

> We can also pass the various options like interactive, fromFile etc for deciding the backend type.

These options are specific to service creation only. For deletion, these options are of no use.

Reply:
> We rely on this information at only one place and there's barely a few lines of code for it:

The same logic is present in create.go. If the deciding logic changes due to some library update then we would need to fix it at both places.

> These options are specific to service creation only. For deletion, these options are of no use.

But it enables us to put the deciding logic in one place.

Reply:
> If the deciding logic changes due to some library update then we would need to fix it at both places.

Makes sense. I'll put this part in one function.

> But it enables us to put the deciding logic in one place. Anyways, we can leave this for now.

This doesn't make sense even for later because:
1. the options are specific only create and not delete
2. they are specific to operators and have nothing to do with service catalog backend.

Reply:
https://github.com/openshift/odo/pull/4465/commits/648eb24cae329b6f9c7ff58cc1192c7112bec25d

Reply:
> the options are specific only create and not delete
they are specific to operators and have nothing to do with service catalog backend.

It puts the decision making in one place and can clean the create.go and delete.go code. Anyways we can skip the code for now.